source/extensions/geoip_providers/maxmind/geoip_provider.cc

```diff
@@ -116,12 +125,48 @@ void GeoipProvider::lookup(Geolocation::LookupRequest&& request,
                            Geolocation::LookupGeoHeadersCallback&& cb) const {
   auto& remote_address = request.remoteAddress();
   auto lookup_result = absl::flat_hash_map<std::string, std::string>{};
-  lookupInCityDb(remote_address, lookup_result);
+  // lookup countryDb only when city_db_path_ is not configured
+  if (!config_->cityDbPath()) {
+    lookupInCountryDb(remote_address, lookup_result);
+  } else {
+    lookupInCityDb(remote_address, lookup_result);
+  }
   lookupInAsnDb(remote_address, lookup_result);
   lookupInAnonDb(remote_address, lookup_result);
   cb(std::move(lookup_result));
 }
 
+void GeoipProvider::lookupInCountryDb(
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    absl::flat_hash_map<std::string, std::string>& lookup_result) const {
```

Comment:
this function has lots of similarity with `lookupInCityDb`. could we refactor to achieve DRY if it's not too hard?

Reply:
there are indeed similarities in these lookup function and ideally be better abstracted, wonder if we could do the refacor in a seperate PR so that this PR focuses on adding country db support

Reply:
let me give it a try, was worried that refactoring here might make this PR complicated to review, maybe could discuss more once there is a draft commit

Reply:
seems one possible approach is to have a common `lookupInMaxmindDb` function, and then we pass the `db_type` and `maxmind_db` database pointer to this function, where `lookupInMaxmindDb` will call related `isLookupEnableForDbType` and `populateGeoLookupResultForDbType` functions, where db-specific logics are handled. 

What do you think about the approach? also still feels like might be better to do in a separate PR to it's cleaner and easier to review (and to revert if ever needed) as other db lookup functions will be touched, thoughts? @nezdolik 

pseudocode
```
void GeoipProvider::lookupInMaxmindDb(
    const Network::Address::InstanceConstSharedPtr& remote_address,
    absl::flat_hash_map<std::string, std::string>& lookup_result,
    const std::string& db_type /* we can make db_type a enum */,
    const std::string& stat_prefix /* e.g. city for city_db */,
    const MaxmindDbPtr& maxmind_db /* passing db pointer */,
    ) const {
  if (isLookupEnableForDbType(db_type))
    ASSERT(maxmind_db, fmt.format("Maxmind {} database is not initialised for performing lookups", stat_prefix));
    int mmdb_error;
    const uint32_t n_prev_hits = lookup_result.size();
    MMDB_lookup_result_s mmdb_lookup_result = MMDB_lookup_sockaddr(
        maxmind_db.get(), reinterpret_cast<const sockaddr*>(remote_address->sockAddr()), &mmdb_error);
    if (!mmdb_error) {
      MMDB_entry_data_list_s* entry_data_list;
      int status = MMDB_get_entry_data_list(&mmdb_lookup_result.entry, &entry_data_list);
      if (status == MMDB_SUCCESS) {
        // populate headers
        populateGeoLookupResultForDbType(db_type)
        if (lookup_result.size() > n_prev_hits) {
          config_->incHit(fmt.format("{}_db", stat_prefix));
        }
        MMDB_free_entry_data_list(entry_data_list);
      }
    } else {
      config_->incLookupError(fmt.format("{}_db", stat_prefix));
    }
    config_->incTotal(fmt.format("{}_db", stat_prefix));
  }
}
```

Reply:
drafted an example PR in https://github.com/envoyproxy/envoy/pull/31490, please take a look when you have a chance